k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/ImapConnection.java

```diff
@@ -326,104 +339,97 @@ private void startTLS() throws IOException, MessagingException, GeneralSecurityE
-                    authXoauth2withSASLIR();
+                    return authXoauth2withSASLIR();
                 } else {
                     throw new MessagingException("Server doesn't support SASL XOAUTH2.");
                 }
-                break;
             case CRAM_MD5: {
                 if (hasCapability(Capabilities.AUTH_CRAM_MD5)) {
-                    authCramMD5();
+                    return authCramMD5();
                 } else {
                     throw new MessagingException("Server doesn't support encrypted passwords using CRAM-MD5.");
                 }
-                break;
             }
             case PLAIN: {
                 if (hasCapability(Capabilities.AUTH_PLAIN)) {
-                    saslAuthPlainWithLoginFallback();
+                    return saslAuthPlainWithLoginFallback();
                 } else if (!hasCapability(Capabilities.LOGINDISABLED)) {
-                    login();
+                    return login();
                 } else {
                     throw new MessagingException("Server doesn't support unencrypted passwords using AUTH=PLAIN " +
                             "and LOGIN is disabled.");
                 }
-                break;
             }
             case EXTERNAL: {
                 if (hasCapability(Capabilities.AUTH_EXTERNAL)) {
-                    saslAuthExternal();
+                    return saslAuthExternal();
                 } else {
                     // Provide notification to user of a problem authenticating using client certificates
                     throw new CertificateValidationException(CertificateValidationException.Reason.MissingCapability);
                 }
-                break;
             }
             default: {
                 throw new MessagingException("Unhandled authentication method found in the server settings (bug).");
             }
         }
     }
 
-    private void authXoauth2withSASLIR() throws IOException, MessagingException {
+    private List<ImapResponse> authXoauth2withSASLIR() throws IOException, MessagingException {
         retryXoauth2WithNewToken = true;
         try {
-            attemptXOAuth2();
+            return attemptXOAuth2();
         } catch (NegativeImapResponseException e) {
             oauthTokenProvider.invalidateToken(settings.getUsername());
 
             if (!retryXoauth2WithNewToken) {
-                handlePermanentXoauth2Failure(e);
+                return handlePermanentXoauth2Failure(e);
             } else {
-                handleTemporaryXoauth2Failure(e);
+                return handleTemporaryXoauth2Failure(e);
             }
         }
     }
 
-    private void handlePermanentXoauth2Failure(NegativeImapResponseException e) throws AuthenticationFailedException {
+    private List<ImapResponse> handlePermanentXoauth2Failure(NegativeImapResponseException e) throws AuthenticationFailedException {
```

Comment:
This feels weird. Maybe make it return `AuthenticationFailedException`, rename to `permanentXoauth2Failure` and change the call sites to

```java
throw permanentXoauth2Failure(…);
```